packages/portalnetwork/src/wire/utp/Protocol/write/ContentWriter.ts

```diff
@@ -29,31 +28,36 @@ export default class ContentWriter {
 
   async start(): Promise<void> {
     const chunks = Object.keys(this.dataChunks).length
-    this.socket.logger(`starting to send ${chunks} DATA Packets`)
     this.writing = true
     let bytes: Uint8Array
-    while (this.writing) {
-      bytes = this.dataChunks[this.seqNr] ?? []
-      this.sentChunks.push(this.seqNr)
-      this.socket.logger(
-        `Sending ST-DATA ${this.sentChunks.length}/${chunks} -- SeqNr: ${this.socket.seqNr}  AckNr: ${this.socket.ackNr}`
-      )
-      await sendDataPacket(this.socket, bytes)
-      this.writing = chunks !== this.sentChunks.length
-      this.seqNr++
+    if (this.sentChunks.length < chunks) {
+      if (this.socket.cur_window + DEFAULT_PACKET_SIZE <= this.socket.max_window) {
+        bytes = this.dataChunks[this.seqNr] ?? []
+        this.sentChunks.push(this.seqNr)
+        this.socket.logger(
+          `Sending ST-DATA ${this.seqNr - this.startingSeqNr + 1}/${chunks} -- SeqNr: ${
+            this.socket.seqNr
+          }`
+        )
+        await sendDataPacket(this.socket, bytes)
+        this.seqNr = Math.max(...this.sentChunks) + 1
```

Comment:
Is it possible these chunks get pushed into the `sentChunks` out of order?  If we're pushing the `seqNr` sequentially into the array, shouldn't we be able to just look at the last one?

Reply:
It could happen if we have to resend a lost packet, and then skip back ahead to where we were.

Preventing the same seqNr from being stored twice is easy though, so unless I'm forgetting something then yes, we should just be able to look at the last seqNr in the array